File: daily_problems/2026/03/0325/personal_submission/cf102556h_yawn_sean.py
# Submission link: https://codeforces.com/gym/102556/submission/368020846
def main(): 
    m = II()
    
    if m == 1:
        exit(print(1))
    
    cnt = Counter()
    
    for x in range(2, 4 * 10 ** 7):
        while m % x == 0:
            cnt[x] += 1
            m //= x
    
    if m > 1:
        cnt[m] += 1
    
    g = 0
    for x in cnt.values():
        g = math.gcd(g, 2 * x)
    
    for i in range(1, g + 1):
        if g % i == 0:
            
            factor_cnt = 1
            for x in cnt:
                factor_cnt *= 1 + 2 * cnt[x] // g * i
            
            if i * factor_cnt == g:
                ans = 1
                for x in cnt:
                    ans *= pow(x, 2 * cnt[x] // g * i)
                print(ans)
                exit()
    
    print(-1)

File: daily_problems/2026/03/0325/personal_submission/cf102556h_yawn_sean.cpp
#include <bits/stdc++.h>
// #pragma GCC optimize("O3,Ofast,unroll-loops")
// #include "atcoder/all"

using namespace std;
auto rng = mt19937(random_device()());
auto rngl = mt19937_64(random_device()());

// Let's unordered_set a bit and flow!
// I came, I divided, I conquered!
// using namespace __gnu_cxx;
// using namespace __gnu_pbds;

int main() {
	ios_base::sync_with_stdio(false);
	cin.tie(0);
	cout.tie(0);

	long long m;
	cin >> m;

	if (m == 1) return cout << 1, 0;

	unordered_map<long long, int> cnt;

	for (int x = 2; x < 40'000'000; x ++) {
		while (m % x == 0) {
			cnt[x] ++;
			m /= x;
		}
	}

	if (m > 1) cnt[m] ++;

	int g = 0;
	for (auto &[k, v]: cnt) g = gcd(g, 2 * v);

	for (int i = 1; i <= g; i ++) {
		if (g % i == 0) {
			int factor_cnt = 1;
			for (auto &[k, v]: cnt)
				factor_cnt *= 1 + 2 * v / g * i;
			if (i * factor_cnt == g) {
				long long ans = 1;
				for (auto &[k, v]: cnt) {
					for (int j = 0; j < 2 * v / g * i; j ++) {
						ans *= k;
					}
				}
				return cout << ans, 0;
			}
		}
	}

	cout << -1;

	return 0;
}